R/render-plot.R

```diff
@@ -106,164 +89,178 @@ renderPlot <- function(expr, width='auto', height='auto', res=72, ...,
+    # Don't invalidate when width/height changes.
+    dims <- if (execOnResize) getDims() else isolate(getDims())
+    pixelratio <- session$clientData$pixelratio %OR% 1
+    hybrid_chain(
+      drawPlot(outputName, session, func, dims$width, dims$height, pixelratio, res),
+      catch = function(reason) {
+        # Non-isolating read. A common reason for errors in plotting is because
+        # the dimensions are too small. By taking a dependency on width/height,
+        # we can try again if the plot output element changes size.
+        getDims()
+
+        # Propagate the error
+        stop(reason)
+      }
+    )
+  })
+
   # This function is the one that's returned from renderPlot(), and gets
-  # wrapped in an observer when the output value is assigned. The expression
-  # passed to renderPlot() is actually run in plotObj(); this function can only
-  # replay a plot if the width/height changes.
+  # wrapped in an observer when the output value is assigned.
   renderFunc <- function(shinysession, name, ...) {
-    session <<- shinysession
     outputName <<- name
+    session <<- shinysession
 
-    dims <- getDims()
-
-    if (is.null(dims$width) || is.null(dims$height) ||
-        dims$width <= 0 || dims$height <= 0) {
-      return(NULL)
-    }
-
-    # The reactive that runs the expr in renderPlot()
-    plotData <- plotObj()
-
-    img <- plotData$img
-
-    # If only the width/height have changed, simply replay the plot and make a
-    # new img.
-    if (dims$width != img$width || dims$height != img$height) {
-      pixelratio <- session$clientData$pixelratio %OR% 1
-
-      coordmap <- NULL
-      plotFunc <- function() {
-        ..stacktraceon..(grDevices::replayPlot(plotData$recordedPlot))
-
-        # Coordmap must be recalculated after replaying plot, because pixel
-        # dimensions will have changed.
-        if (inherits(plotData$plotResult, "ggplot_build_gtable")) {
-          coordmap <<- getGgplotCoordmap(plotData$plotResult, pixelratio, res)
-        } else {
-          coordmap <<- getPrevPlotCoordmap(dims$width, dims$height)
-        }
+    hybrid_chain(
+      drawReactive(),
+      function(result) {
+        dims <- getDims()
+        pixelratio <- session$clientData$pixelratio %OR% 1
+        resizeSavedPlot(name, shinysession, result, dims$width, dims$height, pixelratio, res)
       }
-      outfile <- ..stacktraceoff..(
-        plotPNG(plotFunc, width = dims$width*pixelratio, height = dims$height*pixelratio,
-                res = res*pixelratio)
-      )
-      on.exit(unlink(outfile))
-
-      img <- dropNulls(list(
-        src = session$fileUrl(name, outfile, contentType='image/png'),
-        width = dims$width,
-        height = dims$height,
-        coordmap = coordmap,
-        # Get coordmap error message if present
-        error = attr(coordmap, "error", exact = TRUE)
-      ))
-    }
-
-    img
+    )
   }
 
+  # If renderPlot isn't going to adapt to the height of the div, then the
```

Comment:
Is this related to the async stuff, or is it something that you happened to add on? Is this because of https://github.com/rstudio/shiny/issues/1410 ?

Reply:
It's preexisting: https://github.com/rstudio/shiny/blob/cad20a0bfea9d02dc9f163f028a0c498462fdd90/R/render-plot.R#L263-L264

And it only has to do with Rmd's (where you output `renderPlot(...)` without a corresponding `plotOutput`).